### ISSUE TITLE
IEx open does not work with various editors (e.g. vim) due to assumption of support of `:line`

### ISSUE THREAD

**Issue by aseigo** [elixir-lang/elixir#6346]
### Environment

* Elixir 1.5.-rc.1, OTP 20
* Operating system: Linux

### Current behavior

The new open/1 command in IEx ends up opening the file with a `:<line number>` appended to the file name. This makes vim unhappy ;) It would be nice if there was a way to either define a command line with substitutions for filename and line number, or a way to turn off line numbers in the command.

### Expected behavior

`IEx.open <modulename>` opens the file properly in vim

**Comment by josevalim** [comment]
Is there a variation that works on Vim?
-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Director of R&D

**Comment by sorentwo** [comment]
There is a way to do this directly with vim, but it isn't a human readable "file:line" style:
vim +line file -c 'normal zt'

— Parker

On Mon, Jul 17, 2017, at 01:07 PM, José Valim wrote:
> Is there a variation that works on Vim?
>  -- 
> 
> 
>  *José Valim*
>  www.plataformatec.com.br
>  Skype: jv.ptec
>  Founder and Director of R&D
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub[1], or mute the thread[2]. > 


Links:

  1. https://github.com/elixir-lang/elixir/issues/6346#issuecomment-315834738

**Comment by aseigo** [comment]
Even simpler is possible: vim <filename> +<linenumber>

**Comment by josevalim** [comment]
I have allowed better customization by supporting __FILE__ and __LINE__: d2f52e26dba3aadc4baf9554284f2eeb2eb77666

Please give it a try and let me know if it is not enough.

**Comment by aseigo** [comment]
rock star! :)

thank you very much .. will test the fix a little later.

**Comment by scarfacedeb** [comment]
I can't make `open` to work with vim/nvim in iTerm and macOS. Does it only work in Linux?

I've tried `export ELIXIR_EDITOR="nvim +__LINE__ __FILE__"`, but it didn't help.

It looks like it starts a new vim process, but it's not connected to the current tty.
```
> export ELIXIR_EDITOR="nvim +__LINE__ __FILE__"
> iex
iex(1)> open Tuple.to_list
Vim: Warning: Output is not to a terminal
Vim: Warning: Input is not from a terminal
iex(2)>
```

I noticed that `open` uses `:os.cmd/1` to run the command to open the file. I tried to google it and found the following thread: erlang.2086793.n4.nabble.com/hwo-to-execute-a-text-editor-td4714696.html

I'm not sure if it's the right place to ask the question, but I've tried to ask it in the slack channel already and nobody knew what to do.

**Comment by josevalim** [comment]
Yes, your editor needs to support opening into an existing buffer via a separate program. I know that emacs supports it via the server, other editors like textmate, sublime and atom handle it out of the box as well.

**Comment by josevalim** [comment]
To further clarify, you need to execute a command that will find a running vim instance in the system and then instruct it to open up the file for you. The original command will block until the vim instance that has opened the file closes it. Someone probably has solved this for vim.

**Comment by mhinz** [comment]
@scarfacedeb 

Since you use Neovim, you can do this:

- install [neovim-remote](https://github.com/mhinz/neovim-remote) (provides `nvr` executable)
- start a nvim process with a known server address, eg. `NVIM_LISTEN_ADDRESS=/tmp/nvimsocket nvim`
- replace `nvim` by `nvr`: `export ELIXIR_EDITOR='nvr +__LINE__ __FILE__'`

(Please use the neovim-remote issue tracker for further usage questions.)

**Comment by wojtekmach** [comment]
I'm using terminal vim and I'm currently using the following configuration:

```
export ELIXIR_EDITOR="itermvim __FILE__:__LINE__"
```

```
~% cat bin/itermvim
#!/usr/bin/env bash
osascript -e 'on run argv' -e "tell application \"iTerm2\" to tell current window to create tab with profile \"itermvim\" command \"vi $1\"" -e 'end run' > /dev/null
```

I've also seen folks setting:

```
export ELIXIR_EDITOR="tmux new-window 'vi __FILE__:__LINE__'"
```

**Comment by scarfacedeb** [comment]
@josevalim Thank you for such a quick response. It wasn't clear from the discussion above that it's the only way to `open` modules in vim with iex, and your clarification helped clear things.

@mhinz Thanks! I've actually stumbled upon your dotfiles today and noticed that you're using neovim-remote :) 
I experimented with it just now and even managed to make it work with multiple nvim instances replying only to `iex` inside the same tmux session. Now it's close enough to my usual workflow.

@wojtekmach Another excellent solution, thank you.
